contracts/0.4.24/oracle/LidoOracle.sol

```diff
@@ -73,153 +76,211 @@ contract LidoOracle is ILidoOracle, AragonApp {
-        }
+    /**
+     * @notice Returns the callback contract address to be called upon quorum
+     */
+    function getQuorumCallback() public view returns(address) {
+        return address(QUORUM_CALLBACK_POSITION.getStorageUint256());
+    }
 
-        _assertInvariants();
+    /**
+     * @notice Set the callback contract address to be called upon quorum
+     */
+    function setQuorumCallback(address _addr) external auth(SET_QUORUM_CALLBACK) {
+        QUORUM_CALLBACK_POSITION.setStorageUint256(uint256(_addr));
+        emit QuorumCallbackSet(_addr);
     }
 
     /**
      * @notice An oracle committee member reports data from the ETH 2.0 side
-     * @param _epochId BeaconChain epoch id
+     * @param _epochId Beacon Chain epoch id
      * @param _beaconBalance Balance in wei on the ETH 2.0 side
      * @param _beaconValidators Number of validators visible on this epoch
      */
     function reportBeacon(uint256 _epochId, uint128 _beaconBalance, uint128 _beaconValidators) external {
-        (uint256 startEpoch, uint256 endEpoch) = getCurrentReportableEpochs();
-        require(_epochId >= startEpoch, "EPOCH_IS_TOO_OLD");
+        BeaconSpec memory beaconSpec = _getBeaconSpec();
+        (uint256 currentEpoch, uint256 endEpoch) = getCurrentReportableEpochs();
+        require(_epochId >= currentEpoch, "EPOCH_IS_TOO_OLD");
         require(_epochId <= endEpoch, "EPOCH_HAS_NOT_YET_BEGUN");
+        require(_epochId.mod(beaconSpec.epochsPerFrame) == 0, "MUST_BE_FIRST_EPOCH_FRAME");
 
-        address member = msg.sender;
-        uint256 index = _getMemberId(member);
-        require(index != MEMBER_NOT_FOUND, "MEMBER_NOT_FOUND");
+        // If reported epoch has advanced, clear the last unsuccessful reporting
+        if (_epochId > currentEpoch) _clearReportingAndAdvanceTo(_epochId);
 
-        // check & set contribution flag
-        uint256 bitMask = gatheredEpochData[_epochId].reportsBitMask;
+        // Make sure the oracle is from members list and has not yet voted
+        uint256 index = _getMemberId(msg.sender);
+        require(index != MEMBER_NOT_FOUND, "MEMBER_NOT_FOUND");
+        uint256 bitMask = REPORTS_BITMASK_POSITION.getStorageUint256();
         require(!bitMask.getBit(index), "ALREADY_SUBMITTED");
+        REPORTS_BITMASK_POSITION.setStorageUint256(bitMask.setBit(index, true));
 
-        LAST_REPORTED_EPOCH_ID_POSITION.setStorageUint256(_epochId);
-
-        gatheredEpochData[_epochId].reportsBitMask = bitMask.setBit(index, true);
-
-        Report memory currentReport = Report(_beaconBalance, _beaconValidators);
-        gatheredEpochData[_epochId].reports[index] = currentReport;
-
+        // Push this report to the matching kind
+        uint256 i = 0;
+        Report memory report = Report(_beaconBalance, _beaconValidators);
+        uint256 reportRaw = reportToUint256(report);
+
+        // so iterate on all report kinds we alredy have, not more then oracle members maximum
```

Comment:
If the current report results in a quorum, then the contract does not have to store the passed value in the contract storage. This improvement could save us minimum 5k gas for this tx.

Reply:
There's the downside that it won't be queryable for monitoring purposes though

Reply:
Fix in https://github.com/lidofinance/lido-dao/pull/311/files for clarity.
The last report would not have been seen in monitoring anyway, because it would be cleared immediately after the push to Lido. Good catch. Fixed that first.